examples/door-lock-app/linux/src/LockEndpoint.cpp

```diff
@@ -0,0 +1,339 @@
+    user.userStatus       = userInDb.userStatus;
+    if (DlUserStatus::kAvailable == user.userStatus)
+    {
+        ChipLogDetail(Zcl, "Found unoccupied user [endpoint=%d,adjustedIndex=%hu]", mEndpointId, adjustedUserIndex);
+        return true;
+    }
+
+    user.userName       = chip::CharSpan(userInDb.userName, strlen(userInDb.userName));
+    user.credentials    = chip::Span<const DlCredential>(userInDb.credentials, userInDb.totalCredentials);
+    user.userUniqueId   = userInDb.userUniqueId;
+    user.userType       = userInDb.userType;
+    user.credentialRule = userInDb.credentialRule;
+    user.createdBy      = userInDb.createdBy;
+    user.lastModifiedBy = userInDb.lastModifiedBy;
+
+    ChipLogDetail(
+        Zcl,
+        "Found occupied user "
+        "[endpoint=%d,adjustedIndex=%hu,name=\"%*.s\",credentialsCount=%zu,uniqueId=%x,type=%" PRIu8 ",credentialRule=%" PRIu8 ","
+        "createdBy=%d,lastModifiedBy=%d]",
+        mEndpointId, adjustedUserIndex, static_cast<int>(user.userName.size()), user.userName.data(), user.credentials.size(),
+        user.userUniqueId, to_underlying(user.userType), to_underlying(user.credentialRule), user.createdBy, user.lastModifiedBy);
+
+    return true;
+}
+
+bool LockEndpoint::SetUser(uint16_t userIndex, chip::FabricIndex creator, chip::FabricIndex modifier,
+                           const chip::CharSpan & userName, uint32_t uniqueId, DlUserStatus userStatus, DlUserType usertype,
+                           DlCredentialRule credentialRule, const DlCredential * credentials, size_t totalCredentials)
+{
+    ChipLogProgress(Zcl,
+                    "Door Lock App: LockEndpoint::SetUser "
+                    "[endpoint=%d,userIndex=%" PRIu16 ",creator=%d,modifier=%d,userName=\"%*.s\",uniqueId=%" PRIx32
+                    ",userStatus=%" PRIu8 ",userType=%" PRIu8 ","
+                    "credentialRule=%" PRIu8 ",credentials=%p,totalCredentials=%zu]",
+                    mEndpointId, userIndex, creator, modifier, static_cast<int>(userName.size()), userName.data(), uniqueId,
+                    to_underlying(userStatus), to_underlying(usertype), to_underlying(credentialRule), credentials,
+                    totalCredentials);
+
+    uint16_t adjustedUserIndex = static_cast<uint16_t>(userIndex - 1);
+    if (adjustedUserIndex > mLockUsers.size())
+    {
+        ChipLogError(Zcl, "Cannot set user - index out of range [endpoint=%d,index=%d,adjustedUserIndex=%" PRIu16 "]", mEndpointId,
+                     userIndex, adjustedUserIndex);
+        return false;
+    }
+
+    auto & userInStorage = mLockUsers[adjustedUserIndex];
+
+    if (userName.size() > DOOR_LOCK_MAX_USER_NAME_SIZE)
+    {
+        ChipLogError(Zcl, "Cannot set user - user name is too long [endpoint=%d,index=%d,adjustedUserIndex=%" PRIu16 "]",
+                     mEndpointId, userIndex, adjustedUserIndex);
+        return false;
+    }
+
+    strncpy(userInStorage.userName, userName.data(), userName.size());
```

Comment:
If userName.size() == DOOR_LOCK_MAX_USER_NAME_SIZE (which can happen here), userInStorage.userName will end up not terminated.

In general, please prefer Platform::CopyString to strncpy, but also it seems like we should make userName 1 byte bigger?

Reply:
Thanks for pointing to `chip::Platform::CopyString` method, I didn't know it existed! I've updated the code to use this method instead of `strncpy`. 

Regarding the sizes, there are actually two definitions, both of them are in the door-lock-server.h:
1. [DOOR_LOCK_MAX_USER_NAME_SIZE](https://github.com/project-chip/connectedhomeip/blob/142100859ff50d0deef3e038d3201ed795a64522/src/app/clusters/door-lock-server/door-lock-server.h#L52) which is the number of characters in user name string
2. [DOOR_LOCK_USER_NAME_BUFFER_SIZE](https://github.com/project-chip/connectedhomeip/blob/142100859ff50d0deef3e038d3201ed795a64522/src/app/clusters/door-lock-server/door-lock-server.h#L53-L54) which is the size of the buffer to hold user name string and it is equal to `DOOR_LOCK_MAX_USER_NAME_SIZE + 1`

`userInStorage.userName` has the size of `DOOR_LOCK_USER_NAME_BUFFER_SIZE` and we're [making sure](https://github.com/project-chip/connectedhomeip/pull/13863/files/142100859ff50d0deef3e038d3201ed795a64522#diff-180c2da6fe476bec2cea62bb53146edeedf8be2b850fad880bb513c4b8233f16R95) that the size of provided user name is less than or equal to `DOOR_LOCK_MAX_USER_NAME_SIZE` so there should not be any out-of bounds writes here.